pkg/logs/scheduler/scheduler.go

```diff
@@ -33,19 +41,94 @@ var (
 type Scheduler struct {
 	sources  *logsConfig.LogSources
 	services *service.Services
+
+	// context controls the lifetime of this service
+	ctx context.Context
+
+	// cancel cancels the context
+	cancel context.CancelFunc
+
+	// events from the workloadmeta store
+	events chan workloadmeta.EventBundle
+
+	// services we have added, so that we can remove them again, indexed by ID
+	addedServices map[workloadmeta.EntityID]*service.Service
 }
 
-// CreateScheduler creates the scheduler.
-func CreateScheduler(sources *logsConfig.LogSources, services *service.Services) {
-	adScheduler = &Scheduler{
-		sources:  sources,
-		services: services,
+// NewScheduler creates a scheduler that will report to the given LogSources
+// and Services instances.  The scheduler is automatically started.
+func NewScheduler(sources *logsConfig.LogSources, services *service.Services, wlms workloadmeta.Store) *Scheduler {
+	filter := workloadmeta.NewFilter([]workloadmeta.Kind{workloadmeta.KindContainer}, nil)
+	events := wlms.Subscribe("logs", filter)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	sch := &Scheduler{
+		sources:       sources,
+		services:      services,
+		ctx:           ctx,
+		cancel:        cancel,
+		events:        events,
+		addedServices: make(map[workloadmeta.EntityID]*service.Service),
 	}
+	sch.start()
+	return sch
+}
+
+// CreateScheduler creates the global scheduler (which will subsequently be
+// returned from GetScheduler)
+func CreateScheduler(sources *logsConfig.LogSources, services *service.Services, wlms workloadmeta.Store) {
```

Comment:
nit: not a blocker but the name of this method is a bit confusing from the API perspective (`NewScheduler` + `CreateScheduler` almost imply that they do the same thing).

Reply:
Yeah, I agree -- I'll rename this to CreateGlobalScheduler (and GetScheduler to GetGlobalScheduler).  Probably eventually we should not have a global scheduler :)

Reply:
Oh, I forgot this was merged :)  I'll make a card to track not having a global scheduler.